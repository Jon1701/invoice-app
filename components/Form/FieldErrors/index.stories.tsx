import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import FieldErrors from '.';

export default {
  component: FieldErrors,
} as ComponentMeta<typeof FieldErrors>;

const Template: ComponentStory<typeof FieldErrors> = args => (
  <FieldErrors {...args} />
);

export const OneErrorMessage = Template.bind({});
OneErrorMessage.args = {
  errors: ['invalid field value'],
};

export const MultipleErrorMessages = Template.bind({});
MultipleErrorMessages.args = {
  errors: [
    'invalid field value',
    'value must be a number',
    '"Faith of the Heart" is the greatest opening theme of a Star Trek series',
  ],
};

export const EmptyErrorMessages = Template.bind({});
EmptyErrorMessages.args = {
  errors: [],
};

export const UndefinedErrorMessages = Template.bind({});
UndefinedErrorMessages.args = {
  errors: undefined,
};
