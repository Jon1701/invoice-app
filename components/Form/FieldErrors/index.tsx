import React from 'react';
import styled from 'styled-components';

interface ListProps {
  /**
   * Indicates if more than 1 error message is associated with the form field.
   */
  hasMultipleErrors: boolean;
}

/**
 * Styled list.
 */
const List = styled.ul<ListProps>`
  margin-top: 5px;
  padding-left: ${props => (props.hasMultipleErrors ? '15px' : '0')};

  font-size: 0.9em;
  color: #dc143c;

  // Hide bullets if only one error message is provided.
  list-style-type: ${props => (props.hasMultipleErrors ? 'disc' : 'none')};
`;

interface Props {
  /**
   * Array of error messages to be displayed.
   */
  errors?: Array<string>;
}

/**
 * Displays a list of form field error messages associated with one particular
 * form field.
 */
const FieldErrors: React.FC<Props> = ({ errors }: Props) => {
  // Check if there are errors.
  const hasErrors: boolean = errors !== undefined && errors.length > 0;

  // Don't display component if there are no errors to display.
  if (!hasErrors) {
    return <React.Fragment />;
  }

  // Check if there's more than one error to display.
  const hasMultipleErrors: boolean = (errors && errors.length > 1) || false;

  return (
    <List hasMultipleErrors={hasMultipleErrors}>
      {hasErrors
        ? errors?.map((item: string) => (
            <li key={item.replace(' ', '').toLowerCase()}>{item}</li>
          ))
        : undefined}
    </List>
  );
};

export default FieldErrors;
