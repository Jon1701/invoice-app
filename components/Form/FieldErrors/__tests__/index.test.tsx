import React from 'react';
import { describe, expect, test } from '@jest/globals';
import { render } from '@testing-library/react';

import FieldErrors from '..';

const testCases = [
  { case: 'one error message', errors: ['an unknown error occurred'] },
  {
    case: 'multiple error messages',
    errors: ['activate self-destruct', 'abandon ship', 'ship destroyed'],
  },
  { case: 'no error messages', errors: [] },
  { case: 'undefined', errors: undefined },
];

describe('<FieldErrors />', () => {
  describe.each(testCases)('with case: $case', ({ errors }) => {
    test('should render without issues', () => {
      const node = render(<FieldErrors errors={errors} />);
      expect(node).toMatchSnapshot();
    });
  });
});
