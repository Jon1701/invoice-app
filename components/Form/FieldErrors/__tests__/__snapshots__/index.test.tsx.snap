// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FieldErrors /> with case: multiple error messages should render without issues 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="sc-bcXHqe jMGnKo"
      >
        <li>
          activate self-destruct
        </li>
        <li>
          abandon ship
        </li>
        <li>
          ship destroyed
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="sc-bcXHqe jMGnKo"
    >
      <li>
        activate self-destruct
      </li>
      <li>
        abandon ship
      </li>
      <li>
        ship destroyed
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<FieldErrors /> with case: no error messages should render without issues 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<FieldErrors /> with case: one error message should render without issues 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="sc-bcXHqe fszjvr"
      >
        <li>
          an unknown error occurred
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="sc-bcXHqe fszjvr"
    >
      <li>
        an unknown error occurred
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<FieldErrors /> with case: undefined should render without issues 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
